be/src/storage/persistent_index.h

```diff
@@ -0,0 +1,191 @@
+    Status check_not_exist(size_t n, const uint8_t* keys);
+};
+
+// A persistent primary index contains an in-memory L0 and an on-SSD/NVMe L1,
+// this saves memory usage comparing to the orig all-in-memory implementation.
+// This is a internal class and is intended to be used by PrimaryIndex internally.
+// TODO: code skeleton currently, implementation in future PRs
+//
+// Currently primary index is only modified in TabletUpdates::apply process, it's
+// typical use pattern in apply:
+//   pi.prepare(version)
+//   if (pi.upsert(upsert_keys, values, old_values))
+//   pi.erase(delete_keys, old_values)
+//   pi.commit()
+// If any error occurred between prepare and commit, abort should be called, the
+// index maybe corrupted, currently for simplicity, the whole index is cleared and rebuilt.
+class PersistentIndex {
+public:
+    // |path|: directory that contains index files
+    PersistentIndex(const std::string& path);
+    ~PersistentIndex();
+
+    // if index is loaded
+    bool loaded() const { return (bool)_l0; }
+
+    std::string path() const { return _path; }
+
+    size_t key_size() const { return _key_size; }
+
+    size_t size() const { return _size; }
+
+    EditVersion version() const { return _version; }
+
+    // create new empty index
+    Status create(size_t key_size, const EditVersion& version);
+
+    // load required states from underlying file
+    Status load();
+
+    // start modification with intended version
+    Status prepare(const EditVersion& version);
+
+    // abort modification
+    Status abort();
+
+    // commit modification
+    Status commit();
+
+    // batch index operations
+
+    // batch get
+    // |n|: size of key/value array
+    // |keys|: key array as raw buffer
+    // |values|: value array for return values
+    Status get(size_t n, const uint8_t* keys, IndexValue* values);
+
+    // batch upsert
+    // |n|: size of key/value array
+    // |keys|: key array as raw buffer
+    // |values|: value array
+    // |old_values|: return old values for updates, or set to NullValue for inserts
+    Status upsert(size_t n, const uint8_t* keys, const IndexValue* values, IndexValue* old_values);
+
+    // batch insert, return error if key already exists
+    // |n|: size of key/value array
+    // |keys|: key array as raw buffer
+    // |values|: value array
+    // |check_l1|: also check l1 for insertion consistency(key must not exist previously), may imply heavy IO costs
+    Status insert(size_t n, const uint8_t* keys, const IndexValue* values, bool check_l1);
+
+    // batch erase
+    // |n|: size of key/value array
+    // |keys|: key array as raw buffer
+    // |old_values|: return old values if key exist, or set to NullValue if not
+    Status erase(size_t n, const uint8_t* keys, IndexValue* old_values);
+
+private:
+    // index storage directory
+    std::string _path;
+    size_t _key_size = 0;
+    size_t _size = 0;
+    EditVersion _version;
+    std::unique_ptr<IndexL0> _l0;
```

Comment:
Do we need MVCC to realize concurrent of background compaction and foreground `put` and `get`

Reply:
it takes a more complex design and effort to implement MVCC, so currently, it's not supportted